Add `@source` support

RobinMalfait opened the pull request:
This PR is an umbrella PR where we will add support for the new `@source` directive. This will allow you to add explicit content glob patterns if you want to look for Tailwind classes in other files that are not automatically detected yet.

Right now this is an addition to the existing auto content detection that is automatically enabled in the `@tailwindcss/postcss` and `@tailwindcss/cli` packages. The `@tailwindcss/vite` package doesn't use the auto content detection, but uses the module graph instead.

From an API perspective there is not a lot going on. There are only a few things that you have to know when using the `@source` directive, and you probably already know the rules:

1. You can use multiple `@source` directives if you want.
2. The `@source` accepts a glob pattern so that you can match multiple files at once
3. The pattern is relative to the current file you are in
4. The pattern includes all files it is matching, even git ignored files
   1. The motivation for this is so that you can explicitly point to a `node_modules` folder if you want to look at `node_modules` for whatever reason.
6. Right now we don't support negative globs (starting with a `!`) yet, that will be available in the near future.

Usage example:

```css
/* ./src/input.css */
@import "tailwindcss";
@source "../laravel/resources/views/**/*.blade.php";
@source "../../packages/monorepo-package/**/*.js";
```

It looks like the PR introduced a lot of changes, but this is a side effect of all the other plumbing work we had to do to make this work. For example:

1. We added dedicated integration tests that run on Linux and Windows in CI (just to make sure that all the `path` logic is correct)
2. We Have to make sure that the glob patterns are always correct even if you are using `@import` in your CSS and use `@source` in an imported file. This is because we receive the flattened CSS contents where all `@import`s are inlined.
3. We have to make sure that we also listen for changes in the files that match any of these patterns and trigger a rebuild.

PRs:

- [x] https://github.com/tailwindlabs/tailwindcss/pull/14063
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14085
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14079
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14067
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14076
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14080
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14127
- [x] https://github.com/tailwindlabs/tailwindcss/pull/14135

Once all the PRs are merged, then this umbrella PR can be merged. 

> [!IMPORTANT]  
> Make sure to merge this without rebasing such that each individual PR ends up on the main branch.

RobinMalfait committed:
start of @content support

philipp-spiess committed:
Add PostCSS plugin to fix relative `@content` and `@plugin` paths in `@import`ed files (#14063)

We noticed an issue that happened when handling relative file imports in
the `@plugin` and the upcoming `@content` APIs. The problem arises from
relative files that are inside `@import`ed stylesheets. Take, for
example, the following folder structure:

```css
/* src/index.css */
@import "./dir/index.css";
```
```css
/* src/dir/index.css */
@plugin "../../plugin.ts";
```

It's expected that the path is relative to the CSS file that defined it.
However, right now, we use
[`postcss-import`](https://github.com/postcss/postcss-import) to flatten
the CSS file before running the tailwind build step. This causes these
custom-properties to be inlined in a flat file which removes the
information of which file is being referred:

```css
/* src/flat.css */
@plugin "../../plugin.ts"; /* <- This is now pointing to the wrong file */
```

There are generally two approaches that we can do to solve this:

1. **Handle `@import` flattening inside tailwindcss:** While generally
this would give us more freedom and less dependencies, this would
require some work to get all edge cases right. We need to support
layers/conditional imports and also handle all relative urls for
properties like `background-image`.
2. **Rewrite relative paths as a separate postcss visitor:** The
approach this PR takes is instead to implement a custom postcss plugin
that uses the AST to rewrite relative references inside `@plugin` and
`@content`. This has the benefit of requiring little changes to our
existing APIs. The rule is only enabled for relative references inside
`@plugin` and `@content`, so the surface of this rule is very small.

We can use this plugin inside all three current clients:

- `@tailwindcss/postcss` obviously already uses postcss
- `@tailwindcss/cli` also uses postcss to handle `@import` flattening
- `@tailwindcss/vite` allows us to add custom postcss rules via the CSS
pipeline. There are a few cases that we handle with care (e.g. in vite
you can pass a string to the postcss config which is supposed to load
the config from a file).

To validate the changes, we have added both a list of unit test cases to
the plugin itself as well as verified that all three clients are working
as expected:

- `@tailwindcss/postcss` now has an explicit test for this behavior
- `@tailwindcss/cli` and `@tailwindcss/vite` were manually tested by
updating the vite playground. The CLI was run with `--cwd
playgrounds/vite/ -i ./src/app.css -o foo.css`:
<img width="531" alt="Screenshot 2024-07-29 at 11 35 59"
src="https://github.com/user-attachments/assets/78f0acdc-a46c-4c6c-917a-2916417b1001">

philipp-spiess committed:
Fix Windows tests for new postcss plugin (#14085)

This PR fixes the new `postcss-fix-relative-paths` plugin for Windows
paths. The issue was that we mixed Windows-style path separators from
the absolute file paths with the Posix-style separators from globs. This
caused the `dirname` functions to resolve to the wrong files.

To solve this, we now make the difference very clear by calling the
content a `glob`. For globs, we always expect Posix-style path
separators and for the case of making a glob absolute (by prefixing the
directory), we now convert them into Posix-style explicitly.

This PR also fixes an issue where negative rules (e.g. `!./**/*.ts`)
were not properly rewritten.

---------

Co-authored-by: Jordan Pittman <jordan@cryptica.me>

andrefelipeschulle commented:
Any plans to make it available for testing? Thank you for your incredible work...

RobinMalfait committed:
Add `@content` support to `tailwindcss` (#14079)

This PR adds `@content` support to `tailwindcss`'s core package. We will
handle the `@content` and call the `onContentPath` function when it's
encountered.

The `@tailwindcss/cli`, `@tailwindcss/vite` and `@tailwindcss/postcss`
packages have to implement the `onContentPath` such that the necessary
globs are scanned and watchers should be setup with this information.

Example usage:

```css
@content "../../packages/my-sibling-package/src/components/*.tsx";
```

If you are in a monorepo setup, then you could point to other packages
if you want. Another common use case is for Laravel projects if you want
to point to Laravel blade files since they won't be covered by Vite's
module graph:

```css
/* ./resources/css/app.css */
@content "../views/*.blade.php"
```

Note: all globs are relative to the current file you are in.

RobinMalfait commented:
Hey @andrefelipeschulle, this is currently still in development. But I think we should be able to get a next release out by the end of the week 👍

andrefelipeschulle commented:
@RobinMalfait thanks for answering 👍🏻

RobinMalfait committed:
Add `@content` support to `@tailwindcss/cli` (#14067)

This is a continuation of #14048 where we specifically focus on the
`@tailwindcss/cli`.

This PR optimizes the globs such that we always have a clean `base` and
the `glob` (which is the first dynamic part).

This PR also refactors the CLI such that we can watch files defined by
the `@content` at-rules in the CSS. The `@parcel/watcher` watches from a
base path (not a glob), which means that we have to create a watcher for
each glob's base.

This PR also adds a `scanFiles` to the `scanDir` result, so that we can
verify that a scanned file matches the globs we already know from
`scanDir`.

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
Add `@content` support to `@tailwindcss/vite` (#14076)

This PR makes sure that when you use `@content`, that these files are
also being watched if you are using Vite.

In Vite we rely on the module graph instead, but sometimes it can be
useful to watch files outside of the module graph. E.g.: A Laravel
project, where you are watching `*.blade.php` files.

---------

Co-authored-by: Philipp Spiess <hello@philippspiess.com>

RobinMalfait committed:
Add `@content` support to `@tailwindcss/postcss` (#14080)

This PR adds support for `@content` and makes sure that globs are
registered within PostCSS.

From a configuration perspective nothing changes for the end user. It
does however make sure that `@content` can be used and changes in those
globs will be picked up.

RobinMalfait committed:
Rename `@content` to `@source` (#14127)

This PR renames the `@content` directive to `@source`.

RobinMalfait committed:
update changelog

adamwathan commented on packages/@tailwindcss-cli/src/commands/build/index.ts:
```diff
@@ -253,12 +376,13 @@ function handleImports(
   //   - CSS Import Resolve: https://csstools.github.io/css-import-resolve/
 
-  if (!input.includes('@import')) {
+  if (!input.includes('@import') && !input.includes('@plugin') && !input.includes('@source')) {
```
Can revert this based on our call.

adamwathan commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -2,9 +2,10 @@ import { scanDir } from '@tailwindcss/oxide'
+import postcss, { AtRule, type AcceptedPlugin, type PluginCreator } from 'postcss'
 import postcssImport from 'postcss-import'
 import { compile } from 'tailwindcss'
+import fixRelativePathsPlugin from '../../internal-postcss-fix-relative-paths/src'
```
Use package import instead of relative import.

adamwathan commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -42,120 +43,137 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
-        // Do nothing if neither `@tailwind` nor `@apply` is used
-        if (!hasTailwind && !hasApply) return
+          // Setup the compiler if it doesn't exist yet. This way we can
+          // guarantee a `compile()` function is available.
```
This comment should be `build()` not `compile()` I think.

RobinMalfait committed:
Apply general feedback for `@sources` (#14135)

Part of #14078

This PR applies general feedback to the `@sources` umbrella PR.

This PR includes the following changes:
- Used real classes instead of class-like objects.
- Only handle PostCSS imports when `@import` is used.
- Improve some comments
- Improve naming of `GlobEntry` (`glob.glob` -> `glob.pattern`)
- Use `scanFiles` on the `scanDirResult` as-if it's an incremental scan
(under
  the hood we still do a full scan right now).
- Renamed `content_paths` to `sources`

This is a separate PR just so that we can merge the umbrella PR
un-squashed for each sub-PR. This way all the commits here won't show up
in the `next` branch as-is.

adamwathan commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Add support for `inline` option when defining `@theme` values ([#14095](https://github.com/tailwindlabs/tailwindcss/pull/14095))
 - Add `inert` variant ([#14129](https://github.com/tailwindlabs/tailwindcss/pull/14129))
+- Add `@source` support ([#14078](https://github.com/tailwindlabs/tailwindcss/pull/14078), [14063](https://github.com/tailwindlabs/tailwindcss/pull/14063), [14085](https://github.com/tailwindlabs/tailwindcss/pull/14085), [14079](https://github.com/tailwindlabs/tailwindcss/pull/14079), [14067](https://github.com/tailwindlabs/tailwindcss/pull/14067), [14076](https://github.com/tailwindlabs/tailwind ...
```
Let's just reference the umbrella PR here, this looks ridiculous 😄

RobinMalfait replied:
Haha, just did it because the idea was to merge this umbrella one with a merge commit instead of squashed (so that you see each PR as a separate commit).

But we can squash as well. At the end of the day it's 1 feature (and thousand small sub-features to make things work correctly)

adamwathan commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Add support for `inline` option when defining `@theme` values ([#14095](https://github.com/tailwindlabs/tailwindcss/pull/14095))
 - Add `inert` variant ([#14129](https://github.com/tailwindlabs/tailwindcss/pull/14129))
+- Add `@source` support ([#14078](https://github.com/tailwindlabs/tailwindcss/pull/14078), [14063](https://github.com/tailwindlabs/tailwindcss/pull/14063), [14085](https://github.com/tailwindlabs/tailwindcss/pull/14085), [14079](https://github.com/tailwindlabs/tailwindcss/pull/14079), [14067](https://github.com/tailwindlabs/tailwindcss/pull/14067), [14076](https://github.com/tailwindlabs/tailwind ...
```
```suggestion
- Add `@source` support ([#14078](https://github.com/tailwindlabs/tailwindcss/pull/14078))
```

adamwathan committed:
Update CHANGELOG.md

adamwathan commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Add support for `inline` option when defining `@theme` values ([#14095](https://github.com/tailwindlabs/tailwindcss/pull/14095))
 - Add `inert` variant ([#14129](https://github.com/tailwindlabs/tailwindcss/pull/14129))
+- Add support for explicitly registering content paths with new `@source` at-rule ([#14078](https://github.com/tailwindlabs/tailwindcss/pull/14078))
```
```suggestion
- Add support for explicitly registering content paths using new `@source` at-rule ([#14078](https://github.com/tailwindlabs/tailwindcss/pull/14078))
```

adamwathan committed:
Update CHANGELOG.md
